[clang-repl] Fixing vulnerabilities with respect to orc runtime

anutosh491 opened the pull request:
I was trying to play around with clang-repl out of process on my linux machine and I came across these vulnerabilities.

anutosh491 committed:
Fixing vulnerabilities with respect to orc runtime

anutosh491 commented:
No 3 : (proposing a fix through upcoming commits)

I see that the orc runtime on a linux machine would end up somewhere here (as in my case)
```
(test) anutosh491@vv-nuc:/build/anutosh491/llvm-project/build/lib/clang/22/lib/x86_64-unknown-linux-gnu$ ls
liborc_rt.a
```

But on a mac machine, it might end up here (as in @kr-2003 's case)
```
/Users/abhinav/Desktop/Coding/CERN/llvm-project/build/lib/clang/20/lib/darwin/liborc_rt_osx.a
```

So we have access to these two location in the `getOrcRuntimePath` function 
https://github.com/llvm/llvm-project/blob/8ea447b4c4b007ea1c736f71e569fda621f4fdc8/clang/lib/Interpreter/Interpreter.cpp#L395-L399

`getCompilerRTPath` equates to `<ResourceDir>/lib/<OSLibName>` 
`getRuntimePath` equates to `<ResourceDir>/lib/<triple>/`

And the orc runtime can be present in one of these 2 locations but `getOrcRuntimePath` just searches `getCompilerRTPath` for now !

https://github.com/llvm/llvm-project/blob/8ea447b4c4b007ea1c736f71e569fda621f4fdc8/clang/lib/Interpreter/Interpreter.cpp#L405-L415

We need to update this to prioritise both !

anutosh491 committed:
Improved getOrcRuntimePath function

vgvassilev reviewed (commented):
Can you add tests?

vgvassilev commented on clang/lib/Interpreter/Interpreter.cpp:
```diff
@@ -397,33 +397,48 @@ Interpreter::getOrcRuntimePath(const driver::ToolChain &TC) {
+  if (!CompilerRTPath && !ResourceDir) {
+    return llvm::make_error<llvm::StringError>(
+        "Neither CompilerRT path nor ResourceDir path found",
+        std::error_code());
```
I'd keep this two checks separate and returning proper errors specific to the variable.

anutosh491 committed:
Apply review suggestions

anutosh491 commented on clang/lib/Interpreter/Interpreter.cpp:
```diff
@@ -397,33 +397,48 @@ Interpreter::getOrcRuntimePath(const driver::ToolChain &TC) {
+
+  std::string searched;
+
+  if (CompilerRTPath) {
```
Adressed.

anutosh491 commented:
Addressed the suggestions above. 

Trying to forcefully mess up the path in my case shows me this
```
anutosh491@vv-nuc:/build/anutosh491/llvm-project/build/bin$ ./clang-repl   --oop-executor=$PWD/llvm-jitlink-executor
clang version 22.0.0git (https://github.com/anutosh491/llvm-project.git 81ad8fbc2bb09bae61ed59316468011e4a42cf47)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /build/anutosh491/llvm-project/build/bin
Build config: +unoptimized, +assertions
clang-repl: OrcRuntime library not found in: /build/anutosh491/llvm-project/build/lib/clang/22/lib/linux; /build/anutosh491/llvm-project/build/lib/clang/22/lib/x86_64-unknown-linux-gnu
```
Which is covering both paths as we want.

anutosh491 committed:
formatting

vgvassilev reviewed (approved):
Lgtm!

anutosh491 committed:
Merge branch 'main' into fix_vulnerabilities_with_orcruntime

anutosh491 commented:
Thanks for the review. Merging !
